sys/net/network_layer/ng_sixlowpan/ng_sixlowpan.c

```diff
@@ -0,0 +1,253 @@
+ * @file
+ */
+
+#include "kernel_types.h"
+#include "net/ng_netbase.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "net/ng_sixlowpan.h"
+#include "net/ng_sixlowpan/netif.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+static char _stack[NG_SIXLOWPAN_STACK_SIZE];
+
+/* handles NG_NETAPI_MSG_TYPE_RCV commands */
+void _receive(ng_pktsnip_t *pkt);
+/* handles NG_NETAPI_MSG_TYPE_SND commands */
+void _send(ng_pktsnip_t *pkt);
+/* Main event loop for 6LoWPAN */
+static void *_event_loop(void *args);
+
+kernel_pid_t ng_sixlowpan_init(void)
+{
+    if (_pid > KERNEL_PID_UNDEF) {
+        return _pid;
+    }
+
+    _pid = thread_create(_stack, NG_SIXLOWPAN_STACK_SIZE, NG_SIXLOWPAN_PRIO,
+                         CREATE_STACKTEST, _event_loop, NULL, "6lo");
+
+    return _pid;
+}
+
+void _receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *payload;
+    uint8_t *dispatch;
+    ng_netreg_entry_t *entry;
+
+    LL_SEARCH_SCALAR(pkt, payload, type, NG_NETTYPE_SIXLOWPAN);
+
+    if ((payload == NULL) || (payload->size < 1)) {
+        DEBUG("6lo: Received packet has no 6LoWPAN payload\n");
+        ng_pktbuf_release(pkt);
+    }
+
+    dispatch = payload->data;
+
+    if (dispatch[0] == NG_SIXLOWPAN_UNCOMPRESSED) {
+        ng_pktsnip_t *sixlowpan;
+        DEBUG("6lo: received uncompressed IPv6 packet\n");
+        payload = ng_pktbuf_start_write(payload);
+
+        if (payload == NULL) {
+            DEBUG("6lo: can not get write access on received packet\n");
+#if defined(DEVELHELP) && defined(ENABLE_DEBUG)
+            ng_pktbuf_stats();
+#endif
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        /* packet is uncompressed: just mark and remove the dispatch */
+        sixlowpan = ng_pktbuf_add(payload, payload->data, sizeof(uint8_t),
+                                  NG_NETTYPE_SIXLOWPAN);
+        LL_DELETE(pkt, sixlowpan);
+        ng_pktbuf_release(sixlowpan);
+    }
+    else {
+        DEBUG("6lo: dispatch %02x ... is not supported\n",
+              dispatch[0]);
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    payload->type = NG_NETTYPE_IPV6;
+
+    entry = ng_netreg_lookup(NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL);
+
+    if (entry == NULL) {
+        DEBUG("ipv6: No receivers for this packet found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    ng_pktbuf_hold(pkt, ng_netreg_num(NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL) - 1);
+
+    while (entry) {
+        DEBUG("6lo: Send receive command for %p to %" PRIu16 "\n",
+              (void *)pkt, entry->pid);
+        ng_netapi_receive(entry->pid, pkt);
+        entry = ng_netreg_getnext(entry);
+    }
+}
+
+void _send(ng_pktsnip_t *pkt)
+{
+    ng_netif_hdr_t *hdr;
+    ng_pktsnip_t *ipv6, *sixlowpan;
+    ng_sixlowpan_netif_t *iface;
+    /* cppcheck: datagram_size will be read by frag */
+    /* cppcheck-suppress unreadVariable */
+    size_t payload_len, datagram_size;
```

Comment:
@gebart I know that. The variable is supposed to be used in #2781 and I did not introduce Toi many lines I would revert there anyways.